### PR TITLE
`externalTrafficPolicy` configuration support for Ingress/GwAPI K8s Service

### PR DESCRIPTION
**Pull request by PhilipSchmid**
This PR adds support for configuring the `externalTrafficPolicy` Kubernetes Service spec field of the Cilium Ingress/GatewayAPI `LoadBalancer`/`NodePort` Service.

- Added configuration option to explicitly configure the externalTrafficPolicy for the Cilium Ingress Kubernetes Service.
- Added externalTrafficPolicy support for dedicated Cilium Ingress instances. Configurable via new `ingress.cilium.io/service-external-traffic-policy` Ingress annotation.
- Added externalTrafficPolicy support for Cilium GatewayAPI. eTP is globally configurable via the `gatewayAPI.externalTrafficPolicy` Helm flag.

```release-note
externalTrafficPolicy support for Cilium Ingress and GatewayAPI
```

If possible, let's please try to merge this before the 1.16 feature freeze 🤞.

Open questions based on the outcome of manual E2E testing:

- [x] Changing `ingressController.service.externalTrafficPolicy=Cluster` to `ingressController.service.externalTrafficPolicy=Local` and re-running `make kind-install-cilium` causes the Helm deployment to fail. It looks like it's just a `make kind-install-cilium` "issue" as it runs `cilium install` in the background, and if I manually run `cilium upgrade` with the same parameters, it works. Is this a known behavior? Just double-checking it's not something related to this change.
- [x] Setting an invalid `ingress.cilium.io/service-external-traffic-policy` value on an `ingress.cilium.io/loadbalancer-mode=dedicated` Ingress resource "spams" the Cilium Operator logs with multiple (~10) `Reconciler error` messages. Even after the initial apply, we'll see such an error every other few seconds/minutes. Is there any need to mitigate this? I don't think so, but just checking.
```bash
time="2024-06-04T16:16:39Z" level=info msg="Reconciling Ingress" controller=ingress resource=default/basic-ingress subsys=ingress
time="2024-06-04T16:16:39Z" level=debug msg="Updating dedicated resources" controller=ingress resource=default/basic-ingress subsys=ingress
time="2024-06-04T16:16:39Z" level=error msg="failed to get externalTrafficPolicy annotation from Ingress object" controller=ingress error="failed to parse externalTrafficPolicy \"invalid\"" resource=default/basic-ingress subsys=ingress
time="2024-06-04T16:16:39Z" level=error msg="Reconciler error" Ingress="{basic-ingress default}" controller=ingress controllerGroup=networking.k8s.io controllerKind=Ingress error="failed to build dedicated resources: failed to parse externalTrafficPolicy \"invalid\"" name=basic-ingress namespace=default reconcileID="\"e66ea506-b349-4391-b846-aad5ff30c2f9\"" subsys=controller-runtime
```
- [x] Passing an invalid `gatewayAPI.externalTrafficPolicy` value causes the Cilium Operator to log an error and fail. Does this behavior make sense? I think so, but I'd like to hear other folks' opinions on that.
```bash
time="2024-06-04T14:50:45Z" level=info msg="Cilium Operator 1.16.0-dev eb7b59e954 2024-06-03T16:38:07+02:00 go version go1.22.3 linux/arm64" subsys=cilium-operator-generic
time="2024-06-04T14:50:45Z" level=fatal msg="failed to start: invalid externalTrafficPolicy: invalid-value" subsys=cilium-operator-generic
time=2024-06-04T14:50:45Z level=error msg="Invoke failed" error="invalid externalTrafficPolicy: invalid-value" function="gateway-api.initGatewayAPIController (pkg/gateway-api/cell.go:106)"
time=2024-06-04T14:50:45Z level=info msg=Stopping
```
- [x] It looks like when changing the GwAPI eTP and doing a `helm upgrade`, the eTP on the actual Gateway LB Services is not updated on demand (at least for `<pending>` LB services not). A Cilium Operator restart fixes this and patches the new eTP in the Service.

### CODE REVIEW

File: install/kubernetes/cilium/templates/validate.yaml [hunk 1]
```diff
@@ -60,6 +60,12 @@
   {{- end }}
 {{- end }}
 
+{{- if and .Values.ingressController.enabled (or (eq .Values.ingressController.service.type "LoadBalancer") (eq .Values.ingressController.service.type "NodePort"))}}
```
Comment (mhofstetter) [inline install/kubernetes/cilium/templates/validate.yaml:63]: This validates the ExternalTrafficPolicy Helm value for Ingress - should we do the same for Gateway API?

Referring to one of your questions from the PR description

> Passing an invalid gatewayAPI.externalTrafficPolicy value causes the Cilium Operator to log an error and fail. Does this behavior make sense? I think so, but I'd like to hear other folks' opinions on that.

Already catching this configuration error in the Helm Chart would help preventing the Operator from failing in the first place.

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 2]
```diff
@@ -808,6 +808,13 @@ ingressController:
     # -- Configure if node port allocation is required for LB service
     # ref: https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-nodeport-allocation
     allocateLoadBalancerNodePorts: ~
+    # @schema
```
Comment (mhofstetter) [inline install/kubernetes/cilium/values.yaml.tmpl]: i think we can remove the schema for this value as null isn't allowed? (same as we do below for gateway api.

File: operator/pkg/ingress/annotations/annotations.go [hunk 3]
```diff
@@ -74,6 +75,21 @@ func GetAnnotationServiceType(ingress *networkingv1.Ingress) string {
...
+	val, exists := annotation.Get(ingress, ServiceExternalTrafficPolicyAnnotation)
+	if !exists {
+		return string(corev1.ServiceExternalTrafficPolicyCluster), nil
+	}
+
+	switch val {
+	case string(corev1.ServiceExternalTrafficPolicyCluster), string(corev1.ServiceExternalTrafficPolicyLocal):
+		return val, nil
+	default:
+		return string(corev1.ServiceExternalTrafficPolicyCluster), fmt.Errorf("failed to parse externalTrafficPolicy %q", val)
```
Comment (mhofstetter) [inline operator/pkg/ingress/annotations/annotations.go]: actually it's not a parse error - more an invalid value. Hence, i would rephrase the error accordingly

File: operator/pkg/ingress/ingress_reconcile.go [hunk 4]
```diff
@@ -252,6 +252,11 @@ func (r *ingressReconciler) getSharedListenerPorts() (uint32, uint32, uint32) {
 }
 
 func (r *ingressReconciler) buildDedicatedResources(ctx context.Context, ingress *networkingv1.Ingress) (*ciliumv2.CiliumEnvoyConfig, *corev1.Service, *corev1.Endpoints, error) {
+	scopedLog := r.logger.WithFields(logrus.Fields{
```
Comment (mhofstetter) [inline operator/pkg/ingress/ingress_reconcile.go]: Until no we only logged in the `Reconcile` function - hence defining the scoped logger only there was enough. Not ideal that we have to redefine the scoped logger here again. I wonder whether it would be nicer to pass the scoped logger as additional method parameter (or as part of the context) :thinking:

File: operator/pkg/ingress/ingress_reconcile.go [hunk 5]
```diff
@@ -276,6 +281,13 @@ func (r *ingressReconciler) buildDedicatedResources(ctx context.Context, ingress
 		}
 	}
 
+	eTP, err := annotations.GetAnnotationServiceExternalTrafficPolicy(ingress)
+	if err != nil {
+		scopedLog.WithError(err).Error("failed to get externalTrafficPolicy annotation from Ingress object")
```
Comment (mhofstetter) [inline operator/pkg/ingress/ingress_reconcile.go]: If we only want to warn the user about an invalid annotation value (but continue with the logic) - i suggest to use `Warn` instead of `Error`. 

Referring to one of your questions from the PR description

> Setting an invalid ingress.cilium.io/service-external-traffic-policy value on an ingress.cilium.io/loadbalancer-mode=dedicated Ingress resource "spams" the Cilium Operator logs with multiple (~10) Reconciler error messages. Even after the initial apply, we'll see such an error every other few seconds/minutes. Is there any need to mitigate this? I don't think so, but just checking.

IMO then it's also OK to "spam" the operator log with these warnings.

Note: Log messages should start with Uppercase whereas it's a Go convention to keep errors lowercase -> Change to start uppercase

File: operator/pkg/ingress/ingress_reconcile.go [hunk 6]
```diff
@@ -276,6 +281,13 @@ func (r *ingressReconciler) buildDedicatedResources(ctx context.Context, ingress
...
+	} else {
```
Comment (mhofstetter) [inline operator/pkg/ingress/ingress_reconcile.go]: Else case isn't necessary as `eTP` is set to the default `Cluster` in case the annotation contained an invalid value? Hence it can be always set after checking for the error.

File: operator/pkg/model/translation/gateway-api/translator_test.go [hunk 7]
```diff
@@ -552,14 +554,58 @@ func Test_getService(t *testing.T) {
...
 						},
 					},
-					Type: corev1.ServiceTypeLoadBalancer,
+					Type:                  corev1.ServiceTypeLoadBalancer,
+					ExternalTrafficPolicy: corev1.ServiceExternalTrafficPolicyCluster,
+				},
+			},
+		},
+		{
+			name: "long name - more than 64 characters",
```
Comment (mhofstetter) [inline operator/pkg/model/translation/gateway-api/translator_test.go]: We can adjust the name of this test (and probably change the names of the `Service` in the test accordingly. As the actual test is to set `ExternalTrafficPolicy` to `"Local"`

File: operator/pkg/gateway-api/cell.go [hunk 8]
```diff
@@ -172,6 +180,14 @@ func registerSecretSync(params gatewayAPIParams) secretsync.SecretSyncRegistrati
 	}
 }
 
+func validExternalTrafficPolicy(params gatewayAPIParams) error {
```
Comment (mhofstetter) [inline operator/pkg/gateway-api/cell.go]: nit: `valid` -> `validate`?

File: operator/pkg/gateway-api/cell.go [hunk 9]
```diff
@@ -139,7 +147,7 @@ func initGatewayAPIController(params gatewayAPIParams) error {
 
 	cecTranslator.WithUseAlpn(params.GatewayApiConfig.EnableGatewayAPIAlpn)
 
-	gatewayAPITranslator := gatewayApiTranslation.NewTranslator(cecTranslator, params.GatewayApiConfig.GatewayAPIHostnetworkEnabled)
+	gatewayAPITranslator := gatewayApiTranslation.NewTranslator(cecTranslator, params.GatewayApiConfig.GatewayAPIHostnetworkEnabled, params.GatewayApiConfig.GatewayAPIServiceExternalTrafficPolicy)
```
Comment (mhofstetter) [inline operator/pkg/gateway-api/cell.go]: nit: i suggest to introduce line-breaks between the long parameters the same way as it's done above for the `cecTranslator`

**Review by mhofstetter — CHANGES_REQUESTED** [review]
Thanks for this PR @PhilipSchmid :tada: 

I added some comments inline.

> It looks like when changing the GwAPI eTP and doing a helm upgrade, the eTP on the actual Gateway LB Services is not updated on demand (at least for <pending> LB services not). A Cilium Operator restart fixes this and patches the new eTP in the Service.

Does the Operator gets automatically restarted during the Helm Upgrade (after changing the Helm values)? If not, this would be the issue. (That's why it's working after a manual Operator restart)

> Changing ingressController.service.externalTrafficPolicy=Cluster to ingressController.service.externalTrafficPolicy=Local and re-running make kind-install-cilium causes the Helm deployment to fail. It looks like it's just a make kind-install-cilium "issue" as it runs cilium install in the background, and if I manually run cilium upgrade with the same parameters, it works. Is this a known behavior? Just double-checking it's not something related to this change.

Probably not related to this PR :crossed_fingers: 

-> I tried to address the other questions in my comments inline

**Comment by PhilipSchmid** [comment]
Rebased to `main`

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 2 above)
Comment (gandro) [reply install/kubernetes/cilium/values.yaml.tmpl]: This still seems unresolved

Comment (PhilipSchmid) [reply install/kubernetes/cilium/values.yaml.tmpl]: Yeah, sorry it is. I'm currently finalizing the fixed commits and will commit them in a few minutes. 😁

File: Documentation/helm-values.rst [hunk 10]
```diff
@@ -2231,7 +2231,7 @@
    * - :spelling:ignore:`ingressController.service`
      - Load-balancer service in shared mode. This is a single load-balancer service for all Ingress resources.
      - object
-     - ``{"allocateLoadBalancerNodePorts":null,"annotations":{},"insecureNodePort":null,"labels":{},"loadBalancerClass":null,"loadBalancerIP":null,"name":"cilium-ingress","secureNodePort":null,"type":"LoadBalancer"}``
+     - ``{"allocateLoadBalancerNodePorts":null,"annotations":{},"externalTrafficPolicy":"Cluster","insecureNodePort":null,"labels":{},"loadBalancerClass":null,"loadBalancerIP":null,"name":"cilium-ingress","secureNodePort":null,"type":"LoadBalancer"}``
```
Comment (gandro) [inline Documentation/helm-values.rst:2238]: nit: This should ideally be part of the previous commit

**Comment by PhilipSchmid** [comment]
Thanks for the feedback, @mhofstetter. I just implemented it and force-pushed the fixed commits.

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 2 above)
Comment (PhilipSchmid) [reply install/kubernetes/cilium/values.yaml.tmpl]: Now it should be fixed :).

File: install/kubernetes/cilium/templates/validate.yaml (hunk 1 above)
Comment (PhilipSchmid) [reply install/kubernetes/cilium/templates/validate.yaml:63]: 💯, implemented.

**Review by gandro — APPROVED** [review]
Helm looks good to me (one minor nit)

File: Documentation/helm-values.rst (hunk 10 above)
Comment (PhilipSchmid) [reply Documentation/helm-values.rst:2238]: Good catch, thanks! I addressed this in the newest force-commit.

**Comment by PhilipSchmid** [comment]
@mhofstetter Regarding this 👇

> > It looks like when changing the GwAPI eTP and doing a helm upgrade, the eTP on the actual Gateway LB Services is not updated on demand (at least for LB services not). A Cilium Operator restart fixes this and patches the new eTP in the Service.

> 
> Does the Operator gets automatically restarted during the Helm Upgrade (after changing the Helm values)? If not, this would be the issue. (That's why it's working after a manual Operator restart)

No, the Ciliium Operator isn't automatically restarted, as it's only a `kube-system/cilium-config` CM change. However, the same basically also applies to all of the following GwAPI-related flags: https://github.com/cilium/cilium/blob/4391c8f4a668d7626d0f7a6bbe8318420e98f7e5/install/kubernetes/cilium/templates/cilium-configmap.yaml#L276-L286

For example, if I change `.Values.gatewayAPI.enabled` from `true` to `false` or vice-versa, the Cilium Operator also doesn't get restarted. Hence, if we would like to address this, we would need a generic solution for all "`kube-system/cilium-config` CM only" values.

**Comment by PhilipSchmid** [comment]
I did some manual E2E testing, and it looks good, but I couldn't test it 100%. The main reason for that is that a GwAPI bug was introduced with the commits of https://github.com/cilium/cilium/pull/32744 because with [the commit right before](https://github.com/cilium/cilium/commit/1b987b6bd2434f4b7296a7cd1d031efe75db9160) it still worked. GwAPI engineers are aware of this and provide a fix anytime soon. However, this should not impact or delay this PR.

<details>
<summary>Detailed Instructions for GwAPI on AKS BYOCNI E2E Testing</summary>

# GwAPI on AKS (BYOCNI) E2E Testing

## Cluster Setup
Kube-Proxy Replacement prerequisites:
```bash
az extension add --name aks-preview
az extension update --name aks-preview

az feature register --namespace "Microsoft.ContainerService" --name "KubeProxyConfigurationPreview"
az provider register --namespace Microsoft.ContainerService
```

```bash
set -gx CLUSTERNAME "etp-test-philip"
set -gx RESOURCEGROUP "etp-test-philip"
set -gx LOCATION "eastus2"

az group create --name $RESOURCEGROUP --location $LOCATION

echo >kube-proxy.json '{
  "enabled": false,
  "mode": "IPVS",
  "ipvsConfig": {
      "scheduler": "LeastConnection",
      "TCPTimeoutSeconds": 900,
      "TCPFINTimeoutSeconds": 120,
      "UDPTimeoutSeconds": 300
  }
}'

az aks create --name $CLUSTERNAME --resource-group $RESOURCEGROUP \
    --network-plugin none \
    --kube-proxy-config kube-proxy.json \
    --kubernetes-version 1.29

az aks get-credentials -g $RESOURCEGROUP --name $CLUSTERNAME

kubectl get nodes -o wide
```

## Cilium Installation
Install GwAPI CRDs:
```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/v1.0.0/config/crd/standard/gateway.networking.k8s.io_gatewayclasses.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/v1.0.0/config/crd/standard/gateway.networking.k8s.io_gateways.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/v1.0.0/config/crd/standard/gateway.networking.k8s.io_httproutes.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/v1.0.0/config/crd/standard/gateway.networking.k8s.io_referencegrants.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/v1.0.0/config/crd/experimental/gateway.networking.k8s.io_grpcroutes.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/v1.0.0/config/crd/experimental/gateway.networking.k8s.io_tlsroutes.yaml
```

Install Cilium:
```bash
echo >etp-test-values.yaml "\
image:
  override: "quay.io/cilium/cilium-ci:0879c41b0c952972f781af4c44c14b78b8e07f61@sha256:57d4325ab70d2e6b4de6bc17b17d3af27693b26adf5dd8933e00b403cfdae382"
operator:
  image:
    override: "quay.io/cilium/operator-generic-ci:0879c41b0c952972f781af4c44c14b78b8e07f61@sha256:e0485aecc65d3f0eef8fa67030104a5f5a0cb2d1f2c95f723cd14423433390bf"

kubeProxyReplacement: \"true\"
k8sServiceHost: <TBD>
k8sServicePort: 443

ingressController:
  enabled: true
  loadbalancerMode: shared
  service:
    externalTrafficPolicy: Local
gatewayAPI:
  enabled: true
  externalTrafficPolicy: Local"

cilium install --datapath-mode aks-byocni \
  --chart-directory=$HOME/git/github.com/philipschmid/cilium/install/kubernetes/cilium \
  --helm-values=etp-test-values.yaml

# Wait a few seconds
cilium status
```

## Workload
Test infra workload:
```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: infra
---
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: shared-gateway
  namespace: infra
spec:
  gatewayClassName: cilium
  listeners:
  - name: http
    protocol: HTTP
    port: 80
    allowedRoutes:
      namespaces:
        from: All
```

Get the LB IP/name (wait a few seconds):
```bash
set -gx LB_IP (kubectl get services \
  --namespace infra \
  cilium-gateway-shared-gateway \
  --output jsonpath='{.status.loadBalancer.ingress[0].ip}')
echo $LB_IP
4.153.48.103 # <- IP in my test case
```

Test workload (replace `LB_IP` within the `HTTPRoute`, in my example with `4.153.48.103`):
```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: workload
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: echo-app-deployment
  namespace: workload
  labels:
    app: echo-app
spec:
  replicas: 3 # 3 replicas because my testing AKS cluster has 3 nodes
  selector:
    matchLabels:
      app: echo-app
  template:
    metadata:
      labels:
        app: echo-app
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                  - key: "app"
                    operator: In
                    values:
                    - echo-app
              topologyKey: "kubernetes.io/hostname"
      containers:
      - name: echo-app
        image: ghcr.io/philipschmid/echo-app:main
        ports:
        - containerPort: 8080
        env:
        - name: PRINT_HTTP_REQUEST_HEADERS
          value: "true"
---
apiVersion: v1
kind: Service
metadata:
  name: echo-app-service
  namespace: workload
  labels:
    app: echo-app
spec:
  selector:
    app: echo-app
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 8080
  type: ClusterIP
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: workload
  namespace: workload
spec:
  parentRefs:
  - name: shared-gateway
    namespace: infra
  hostnames:
    - gwapi.$LB_IP.sslip.io # gwapi.4.153.48.103.sslip.io
  rules:
  - backendRefs:
    - name: echo-app-service
      port: 8080
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: workload
  namespace: workload
spec:
  ingressClassName: cilium
  rules:
  - host: ingress.$LB_IP.sslip.io # ingress.4.153.48.103.sslip.io
    http:
      paths:
      - pathType: Prefix
        path: "/"
        backend:
          service:
            name: echo-app-service
            port:
              number: 8080
```

## Testing
Test GwAPI:
```bash
curl http://gwapi.4.153.48.103.sslip.io
```

Test Ingress:
```bash
curl http://ingress.4.153.48.103.sslip.io
```

# Clean Up
```bash
az aks delete --name $CLUSTERNAME --resource-group $RESOURCEGROUP
az group delete --name $RESOURCEGROUP
```

</details>

**Review by qmonnet — APPROVED** [review]
Doc changes look good, thanks!

A few (non-blocking) observations on the commit formatting:
- 3rd commit description has a typo: globaly -> globally
- 3rd commit title (`gwapi: eTP support for GwAPI`) is also not super easy to read because of the many abbreviations, I'd consider expanding at least one of the two and using consistent case on gwapi/GwAPI if you keep it abbreviated
- We usually wrap commit descriptions on 72-character wide lines (see most commit descriptions in `git log`)
- On your way to become the first `@cisco.com` commit author! :sunglasses:

**Comment by PhilipSchmid** [comment]
Thanks a lot for the input, @qmonnet! I just reworded the commit messages and force-pushed once again.

**Review by mhofstetter — APPROVED** [review]
LGTM - great work @PhilipSchmid :tada:
